dsls/sql/src/main/java/org/apache/beam/dsls/sql/BeamSql.java

```diff
@@ -93,42 +98,62 @@
    *
    * <p>Make sure to query it from a static table name <em>PCOLLECTION</em>.
    */
-  public static PTransform<PCollection<BeamSqlRow>, PCollection<BeamSqlRow>>
-  simpleQuery(String sqlQuery) throws Exception {
-    return new SimpleQueryTransform(sqlQuery);
+  public static SimpleQueryTransform simpleQuery(String sqlQuery) throws Exception {
+    return SimpleQueryTransform.builder()
+        .setSqlEnv(new BeamSqlEnv())
+        .setSqlQuery(sqlQuery)
+        .build();
   }
 
   /**
    * A {@link PTransform} representing an execution plan for a SQL query.
    */
-  private static class QueryTransform extends
+  @AutoValue
+  public abstract static class QueryTransform extends
       PTransform<PCollectionTuple, PCollection<BeamSqlRow>> {
-    private transient BeamSqlEnv sqlEnv;
-    private String sqlQuery;
+    abstract BeamSqlEnv getSqlEnv();
+    abstract String getSqlQuery();
 
-    public QueryTransform(String sqlQuery) {
-      this.sqlQuery = sqlQuery;
-      sqlEnv = new BeamSqlEnv();
+    static Builder builder() {
+      return new AutoValue_BeamSql_QueryTransform.Builder();
     }
 
-    public QueryTransform(String sqlQuery, BeamSqlEnv sqlEnv) {
-      this.sqlQuery = sqlQuery;
-      this.sqlEnv = sqlEnv;
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setSqlQuery(String sqlQuery);
+      abstract Builder setSqlEnv(BeamSqlEnv sqlEnv);
+      abstract QueryTransform build();
     }
 
+    /**
+     * register a UDF function used in this query.
+     */
+     public QueryTransform withUdf(String functionName, Class<?> clazz, String methodName){
```

Comment:
1. Okay
2. We may want to revisit this when we do a holistic API review before merge to master, but fine for now then.
3. That's true, but for Java 8 users it'd be nice to have an overload supporting a lambda. Not necessary for this PR though.